File: Chapter 10/Python/section_10_3_resilience_decorators/resilient/core/ports/payment_gateway.py
from abc import ABC, abstractmethod

class PaymentGateway(ABC):
    """
    THE CORE PORT (The Primary Airlock):
    
    DESIGN NOTE:
    This Port is the Core's definition of how it expects to interact with the 
    physical world. It belongs to the Core, not the Infrastructure.
    
    ARCHITECTURAL CRITIQUE:
    If this interface were in 'infrastructure', the Core would depend on 
    Infrastructure—violating the Golden Rule. By placing this in 'core/ports', 
    we force the external adapters to point INWARD to satisfy our business needs.
    """
    @abstractmethod
    def charge(self, amount: float, order_id: str, idempotency_key: str) -> bool:
        pass

File: Chapter 10/JS/section_10_3_resilience_decorators/resilient/core/ports/paymentGateway.js
/**
 * THE CORE PORT (The Primary Airlock):
 * * @interface
 * @description
 * Since Vanilla JS doesn't have native Interfaces, we provide a 
 * "Contract Template." Any adapter must implement a 'charge' method.
 * * ARCHITECTURAL CRITIQUE:
 * By defining this Port in the Core, we invert the dependency. 
 * Infrastructure now depends on the Core's requirements. This is 
 * the heart of a "Clarity Engineer's" boundary management.
 */
export class PaymentGateway {
    /**
     * @param {number} amount 
     * @param {string} orderId 
     * @param {string} idempotencyKey 
     * @returns {Promise<boolean>}
     */
    async charge(amount, orderId, idempotencyKey) {
        throw new Error("Port Method 'charge' must be implemented by an Adapter.");
    }
}